pip/req/req_set.py

```diff
@@ -238,6 +238,15 @@ def add_requirement(self, install_req, parent_req_name=None):
                 existing_req = None
             if (parent_req_name is None and existing_req and not
                     existing_req.constraint):
+                if existing_req.name == install_req.name:
+                    # If the extras do not match, then union them together to
+                    # support install bar bar[foo]. If they do match, the
+                    # InstallationError will be raised as per normal.
+                    if existing_req.extras != install_req.extras:
+                        existing_req.extras = tuple(
+                            set(existing_req.extras).union(
+                                set(install_req.extras)))
+                        return [existing_req]
```

Comment:
I agree it needs to apply whether constraint or not. A new test case is needed to demonstrate this.

@xavfernandez yes, it will give you the union of extras without cross-checking all the version details - that is however consistent with all the other handling we do of conflicting specifiers. I think its better than ignoring the extras entirely as we do today, and is a stepping stone to the full resolver scenario.

I think a worthwhile thing to do would be an expected failure test showing that this should error (as should many other conflicting specifier cases) and we can make sure we cater for this when the resolver branch matures.